base/shell/explorer/trayntfy.cpp

```diff
@@ -33,6 +34,253 @@ typedef struct _SYS_PAGER_COPY_DATA
     NOTIFYICONDATA  nicon_data;
 } SYS_PAGER_COPY_DATA, *PSYS_PAGER_COPY_DATA;
 
+
+struct IconWatcherData
+{
+    HANDLE hProcess;
+    DWORD ProcessId;
+    NOTIFYICONDATA IconData;
+
+    IconWatcherData()
+    {
+        IconWatcherData(NULL);
+    }
+
+    IconWatcherData(NOTIFYICONDATA *iconData) :
+        hProcess(NULL), ProcessId(0)
+    {
+        IconData.cbSize = sizeof(NOTIFYICONDATA);
+        IconData.hWnd = iconData->hWnd;
+        IconData.uID = iconData->uID;
+        IconData.guidItem = iconData->guidItem;
+    }
+
+    ~IconWatcherData()
+    {
+        if (hProcess)
+        {
+            CloseHandle(hProcess);
+        }
+    }
+};
+
+class CIconWatcher
+{
+    CAtlList<IconWatcherData *> m_WatcherList;
+    CRITICAL_SECTION m_ListLock;
+    HANDLE m_hWatcherThread;
+    HANDLE m_WakeUpEvent;
+    HWND m_hwndSysTray;
+    bool m_Loop;
+
+public:
+    CIconWatcher() :
+        m_hWatcherThread(NULL),
+        m_WakeUpEvent(NULL),
+        m_hwndSysTray(NULL),
+        m_Loop(false)
+    {
+    }
+    virtual ~CIconWatcher()
+    {
+        Uninitialize();
+        if (m_WakeUpEvent)
+            CloseHandle(m_WakeUpEvent);
+        if (m_hWatcherThread)
+            CloseHandle(m_hWatcherThread);
+    }
+
+    bool Initialize(_In_ HWND hWndParent)
+    {
+        m_hwndSysTray = hWndParent;
+
+        InitializeCriticalSection(&m_ListLock);
+        m_WakeUpEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
+        if (m_WakeUpEvent == NULL)
+            return false;
+
+        m_hWatcherThread = (HANDLE)_beginthreadex(NULL,
```

Comment:
Are you using "_beginthreadex" instead of CreateThread, because you're going to use some "heavy" CRT functions inside the thread?

Reply:
I always prefer _beginthreadex over CreateThread in applications, especially in C++. There's very little overhead in doing so, and it ensures the thread is correctly setup with the C runtime for anything that needs it.

Reply:
@HBelusca It's not perse for 'heavy' functions, but in general for any CRT functions to have a defined environment.